### PR TITLE
Improve type safety in ObjectIndex implementations

### PR DESCRIPTION
**Pull request by elprans**
Currently, `ObjectIndex` keys are strings, but are supposed to be
`Name` to preserve the name structure.  This happened to touch a bit more
than just names, as I had to fix some of the AST annotations to make code
areas reachable for mypy.

### CODE REVIEW

File: edb/edgeql/compiler/setgen.py [hunk 1]
```diff
@@ -586,6 +591,7 @@ def resolve_ptr(
     track_ref: Optional[Union[qlast.Base, Literal[False]]],
     ctx: context.ContextLevel,
 ) -> s_pointers.Pointer:
+    assert isinstance(pointer_name, str)
```
Comment (msullivan) [inline edb/edgeql/compiler/setgen.py:594]: Why this? Catching some untyped bad callsites?

File: edb/ir/ast.py [hunk 2]
```diff
@@ -420,8 +420,8 @@ class Set(Base):
     path_id: PathId
     path_scope_id: typing.Optional[int]
     typeref: TypeRef
-    expr: Expr
-    rptr: Pointer
+    expr: typing.Optional[Expr]
```
Comment (msullivan) [inline edb/ir/ast.py:423]: Wooo

File: edb/pgsql/metaschema.py [hunk 3]
```diff
@@ -2916,13 +2925,18 @@ def _generate_database_views(schema):
...
-    Role = schema.get('sys::Role')
-    member_of = Role.getptr(schema, 'member_of')
-    bases = Role.getptr(schema, 'bases')
-    ancestors = Role.getptr(schema, 'ancestors')
-    annos = Role.getptr(schema, 'annotations')
-    int_annos = Role.getptr(schema, 'annotations__internal')
+def _generate_role_views(schema: s_schema.Schema) -> List[dbops.View]:
+    Role = schema.get('sys::Role', type=s_objtypes.ObjectType)
+    member_of = Role.getptr(schema, s_name.UnqualName('member_of'))
+    assert isinstance(member_of, s_links.Link)
```
Comment (msullivan) [inline edb/pgsql/metaschema.py:2931]: Would it be worth adding a `type` argument to `getptr`?

**Review by msullivan — APPROVED** [review]

File: edb/edgeql/compiler/setgen.py (hunk 1 above)
Comment (elprans) [reply edb/edgeql/compiler/setgen.py:594]: Yeah.  Not necessary anymore.

File: edb/pgsql/metaschema.py (hunk 3 above)
Comment (elprans) [reply edb/pgsql/metaschema.py:2931]: Yeah, that'd be better.

**Pull request merged**